retis/src/process/cli/pcap.rs

```diff
@@ -68,76 +79,82 @@ macro_rules! some_or_return {
     };
 }
 
+// TODO: Register with IANA?
+const RETIS_PEN: u32 = 70000;
+
+/// Custom block containing the JSON-Schema of events.
+struct SchemaBlock {
+    schema: Schema,
+}
+
+impl SchemaBlock {
+    fn new() -> Result<Self> {
+        Ok(SchemaBlock {
+            schema: schema_for!(Event),
+        })
+    }
+}
+
+impl CustomCopiable<'_> for SchemaBlock {
+    const PEN: u32 = RETIS_PEN;
+    type FromSliceError = SchemaBlockError;
+    type WriteToError = SchemaBlockError;
+
+    fn write_to<W: Write>(&self, writer: &mut W) -> Result<(), SchemaBlockError> {
+        let bytes = serde_json::to_vec(&self.schema)?;
+        writer.write_all(&bytes[..])?;
+        Ok(())
+    }
+
+    fn from_slice(slice: &[u8]) -> Result<Option<Self>, SchemaBlockError> {
+        let schema: Schema = serde_json::from_slice(slice)?;
+        Ok(Some(SchemaBlock { schema }))
+    }
+}
+
+// Custom error used for pcap-file API.
+#[derive(thiserror::Error, Debug)]
+enum SchemaBlockError {
+    #[error("json error: {0}")]
+    Json(#[from] serde_json::Error),
+    #[error("i/o error: {0}")]
+    Io(#[from] std::io::Error),
+}
+
 impl EventParser {
     /// Creates a new EventParser from a PcapNgWriter<W: Write>.
     fn new() -> Self {
         Self {
             ifaces: HashMap::new(),
             stats: EventParserStats::default(),
             ts_off: None,
+            wrote_header: false,
         }
     }
 
-    /// Parse & process a single Retis event.
-    fn parse(&mut self, event: &Event) -> Result<Vec<Block<'_>>> {
-        // Having a common & a kernel section is mandatory for now, seeing a
-        // filtered event w/o one of those is bogus.
-        let common = event
-            .common
-            .as_ref()
-            .ok_or_else(|| anyhow!("No common section in event"))?;
-        let kernel = event
-            .kernel
-            .as_ref()
-            .ok_or_else(|| anyhow!("No skb section in event"))?;
-
-        self.stats.processed += 1;
-
-        // The packet section is mandatory for us to generate PCAP events, but
-        // it might not be present in some filtered events. Stats are kept to
-        // inform the user.
-        let packet = some_or_return!(&event.packet, self.stats.missing_packet);
-
-        // The dev & ns sections are best to have but not mandatory to generate
-        // an event. If not found, fake them.
-        let (ifindex, ifname) = match &event.dev {
-            Some(dev) => (dev.ifindex, dev.name.as_str()),
-            None => {
-                self.stats.missing_dev += 1;
```

Comment:
`missing_dev`, `missing_skb` and `missing_ns` are not used anymore, they should be removed from `EventParserStats`.